File: src/main/java/pl/yalgrin/playnite/simplesync/domain/PlatformDiff.java
package pl.yalgrin.playnite.simplesync.domain;

import io.r2dbc.postgresql.codec.Json;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.experimental.SuperBuilder;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.Transient;
import org.springframework.data.relational.core.mapping.Column;
import org.springframework.data.relational.core.mapping.Table;
import pl.yalgrin.playnite.simplesync.dto.PlatformDiffDTO;

@EqualsAndHashCode(callSuper = true)
@Data
@NoArgsConstructor
@AllArgsConstructor
@SuperBuilder
@Table("playnite_platform_diff")
public class PlatformDiff extends AbstractObjectDiffEntity {
    @Id
    @Column("id")
    private Long id;

    @Column("playnite_id")
    private String playniteId;

    @Column("name")
    private String name;

    @Column("contents")
    private Json diffData;

    @Column("removed")
    private boolean removed;

    @Transient
    private boolean notifyAll;

    @Transient
    private boolean changed;

    @Transient
    private PlatformDiffDTO diffDTO;
}

File: src/main/kotlin/pl/yalgrin/playnite/simplesync/library/domain/PlatformDiff.kt
package pl.yalgrin.playnite.simplesync.library.domain

import io.r2dbc.postgresql.codec.Json
import org.springframework.data.annotation.Id
import org.springframework.data.annotation.Transient
import org.springframework.data.relational.core.mapping.Column
import org.springframework.data.relational.core.mapping.Table
import pl.yalgrin.playnite.simplesync.common.util.asObject
import pl.yalgrin.playnite.simplesync.library.dto.PlatformDiffDatabaseModel
import reactor.core.publisher.Mono
import java.time.Instant

object PlatformDiffConstants {
    const val CURRENT_MODEL_VERSION = 2L
}

@Table("playnite_platform_diff")
data class PlatformDiff(
    @Id
    @Column("id")
    override var id: Long? = null,

    @Column("playnite_id")
    override var playniteId: String? = null,

    @Column("name")
    override var name: String? = null,

    @Column("contents")
    override var diffData: Json? = null,

    @Column("model_version")
    var modelVersion: Long = PlatformDiffConstants.CURRENT_MODEL_VERSION,

    @Column("removed")
    override var isRemoved: Boolean = false,

    @Column("created_at")
    override var createdAt: Instant = Instant.now(),

    @Column("created_by")
    override var createdBy: String? = null,

    @Transient
    var isNotifyAll: Boolean = false,

    @Transient
    var isChanged: Boolean = false,

    @Transient
    override var isForEntireObject: Boolean = false
) : LibraryObjectDiffEntity

fun PlatformDiff.extractDbModelOrEmpty(): Mono<PlatformDiffDatabaseModel> {
    return this.diffData.asObject(PlatformDiffDatabaseModel::class.java)
        .switchIfEmpty(Mono.fromSupplier { PlatformDiffDatabaseModel() })
}